File: src/main/java/com/carlsonwagonlit/linedef/LineDefRepo.java
package com.carlsonwagonlit.linedef;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Repository;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Optional;

/**
 * Repository for the LineDefs defined in the JSON file
 */
@Repository
public class LineDefRepo {
    private static final TypeReference<ArrayList<LineDef>> TYPE_LIST_LINEDEF = new TypeReference<ArrayList<LineDef>>() {
    };
    private static final ObjectMapper mapper = new ObjectMapper();
    private final Logger LOG = LoggerFactory.getLogger(getClass());

    @Value("classpath:linedefs.json")
    private URL lineDefResource;

    private ArrayList<LineDef> lineDefs;
    private LinkedHashMap<LineDefId, LineDef> lineDefMap;

    @PostConstruct
    private void initialize() throws IOException {
//        URL resource = Resources.getResource("linedefs.json");
//        String string = Resources.toString(resource, Charsets.UTF_8);
        //        File file = new ClassPathResource("linedefs.json").getFile();

        lineDefs = mapper.readValue(lineDefResource, TYPE_LIST_LINEDEF);
        LOG.info("lineDefs=" + lineDefs);

        lineDefMap = new LinkedHashMap<>(lineDefs.size());
        for (LineDef lineDef : lineDefs) {
            LineDefId id = new LineDefId(lineDef.getSubGuidId(), lineDef.getGdsId());
            lineDefMap.put(id, lineDef);
        }
    }

    public List<LineDef> lineDefs() {
        return Collections.unmodifiableList(lineDefs);
    }

    public Optional<LineDef> lineDef(String subGuidId, String gdsId) {
        LineDef lineDef = lineDefMap.get(new LineDefId(subGuidId, gdsId));
        return Optional.ofNullable(lineDef);
    }

    private static class LineDefId {
        private final String subGuidId;
        private final String gdsId;

        private LineDefId(String subGuidId, String gdsId) {
            this.subGuidId = subGuidId;
            this.gdsId = gdsId;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            LineDefId lineDefId = (LineDefId) o;

            if (!subGuidId.equals(lineDefId.subGuidId)) return false;
            return gdsId.equals(lineDefId.gdsId);
        }

        @Override
        public int hashCode() {
            int result = subGuidId.hashCode();
            result = 31 * result + gdsId.hashCode();
            return result;
        }
    }
}

File: src/main/kotlin/com/carlsonwagonlit/linedef/LineDefRepo.kt
package com.carlsonwagonlit.linedef

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Repository
import java.net.URL
import java.util.Collections
import java.util.Optional


/**
 * Repository for the LineDefs defined in the JSON file
 */
@Repository
class LineDefRepo(@Value("classpath:linedefs.json") private val lineDefResource: URL) {
    private val LOG = LoggerFactory.getLogger(javaClass)

    private val lineDefMap = initLineDefMap()

    private fun initLineDefMap(): Map<LineDefId, LineDef> {
        val typeRef = object : TypeReference<List<LineDef>>() {}
        val mapper = ObjectMapper()

        // If we want the flexibility to add new fields at the cost of compile-time check:
        // 1) remove unit test "rejectUnknownKey"
        // 2) uncomment line below
//        mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)

        val lineDefs: List<LineDef> = mapper.readValue(lineDefResource, typeRef)

        LOG.info("lineDefs=" + lineDefs)

        val map = LinkedHashMap<LineDefId, LineDef>(lineDefs.size)
        lineDefs.forEach {
            require(it.topGuidId.isNotBlank() && it.subGuidId.isNotBlank() && it.gdsId.isNotBlank())

            val id = LineDefId(it.topGuidId, it.subGuidId, it.gdsId)
            val prev = map.put(id, it)
            if (prev != null) {
                throw IllegalArgumentException("Duplicate LineDef found, id=$id")
            }
            val sourceIds = it.fields.map { it.sourceId }
            if (sourceIds.any(String::isBlank)) {
                throw IllegalArgumentException("LineDef id=$id has field with empty 'sourceId'")
            }
            if (sourceIds.size != HashSet(sourceIds).size) {
                throw IllegalArgumentException("LineDef id=$id has a duplicate sourceId")
            }
        }

        LOG.info("LineDefs parsed=${map.size}")

        return Collections.unmodifiableMap(map)
    }

    fun size(): Int = lineDefMap.size

    fun lineDef(topGuidId: String, subGuidId: String, gdsId: String): Optional<LineDef> {
        val id = LineDefId(topGuidId, subGuidId, gdsId)
        return Optional.ofNullable(lineDefMap[id])
    }

    private data class LineDefId(val topGuidId: String, val subGuidId: String, val gdsId: String)
}

